src/device/wg_controller.rs

```diff
@@ -250,9 +250,19 @@ async fn consolidate_wg_peers<
                     requested_peer.peer.endpoint,
                     requested_peer.local_direct_endpoint,
                 ) {
-                    if let Some(us) = upgrade_sync {
-                        us.request_upgrade(&public_key, remote_endpoint, local_direct_endpoint)
-                            .await?;
+                    if let (Some(us), Some(cpc)) = (upgrade_sync, cross_ping_check) {
+                        match cpc.session_for_key(public_key).await? {
```

Comment:
I wonder if this is really correct.

As `session_for_key()` returns any session which matches provided `public_key`. In general CPC module creates this matrix of:
```
+----------------+------------+------------+------------+
| .              | Peer 1     | Peer 2     | Peer 3     |
+----------------+------------+------------+------------+
| EP Candidate 1 | session 11 | session 12 | session 13 |
+----------------+------------+------------+------------+
| EP Candidate 2 | session 21 | session 22 | session 23 |
+----------------+------------+------------+------------+
| EP Candidate 3 | session 31 | session 32 | session 33 |
+----------------+------------+------------+------------+
```

I think `session_for_key()` function should match not only public key, but also either `local_direct_endpoint`.

Reply:
I probably could have mentioned this without using ASCII tables :D but who can pass such an opportunity :hugs:

:laughing: :laughing: :laughing:

Reply:
Additionally there is possibility to add session ID to [WireGuardEndpointCandidateChangeEvent](https://github.com/NordSecurity/libtelio/blob/main/crates/telio-traversal/src/connectivity_check/mod.rs#L35).

As this gets probed from CPC in wg_controller [here](https://github.com/NordSecurity/libtelio/blob/main/src/device/wg_controller.rs#L649), later passed to [select_endpoint_for_peer()](https://github.com/NordSecurity/libtelio/blob/main/src/device/wg_controller.rs#L701), which depending on conditions may select CPC [provided endpoint](https://github.com/NordSecurity/libtelio/blob/main/src/device/wg_controller.rs#L848), which uses previously mentioned [struct](https://github.com/NordSecurity/libtelio/blob/main/src/device/wg_controller.rs#L784). 

Of course, this would mean, that local_direct_endpoint, would have to carry "session ID" back up to the point where `update_request` is initiated. But it should be quite convenient to do via the [ReqestedPeer](https://github.com/NordSecurity/libtelio/blob/main/src/device/wg_controller.rs#L784) struct. This would allow us to avoid two separate lookups in CPC state. Which, perfromance-wise does not have any impact, but having single "pass" from CPC to WG of the map of [WireGuardEndpointCandidateChangeEvent](https://github.com/NordSecurity/libtelio/blob/main/crates/telio-traversal/src/connectivity_check/mod.rs#L35) structs -> ensures consistency in a trivial manner :).

Reply:
Done